File: lib/core/services/service_area_catalog.dart
import 'package:flutter/foundation.dart';
import 'package:hilla_ride/core/constants/babil_regions.dart';
import 'package:hilla_ride/core/models/service_area_models.dart';
import 'package:hilla_ride/core/utils/geo_polygon.dart';
import 'package:latlong2/latlong.dart';

/// In-memory geo catalog synced from Firestore.
///
/// After the first snapshot, seed data is never used — an empty active set
/// means no new rides (deactivated areas stay out immediately).
class ServiceAreaCatalog extends ChangeNotifier {
  ServiceAreaCatalog._();
  static final ServiceAreaCatalog instance = ServiceAreaCatalog._();

  static const _seedProvinceId = 'babil';
  static const _seedCountryId = 'iq';

  List<ServiceDistrictNode>? _liveDistricts;
  /// All districts (active + inactive) for admin filters / labels.
  List<ServiceDistrictNode>? _allDistricts;
  List<ServiceDistrict> _rawDistricts = const [];
  List<ServiceSubDistrict> _rawSubs = const [];
  Map<String, ServiceSubDistrict> _subById = {};
  Map<String, ServiceProvince> _provincesById = {};
  Map<String, ServiceCountry> _countriesById = {};
  var _synced = false;

  /// True once at least one Firestore snapshot has been applied.
  bool get hasSynced => _synced;

  /// True when live active districts are available for booking.
  bool get hasLiveData =>
      _synced && _liveDistricts != null && _liveDistricts!.isNotEmpty;

  void replaceFromFirestore({
    required List<ServiceCountry> countries,
    required List<ServiceProvince> provinces,
    required List<ServiceDistrict> districts,
    required List<ServiceSubDistrict> subs,
  }) {
    _synced = true;
    _countriesById = {for (final c in countries) c.id: c};
    _provincesById = {for (final p in provinces) p.id: p};
    _rawDistricts = List<ServiceDistrict>.from(districts);
    _rawSubs = List<ServiceSubDistrict>.from(subs);

    bool parentActive(ServiceDistrict d) {
      final province = _provincesById[d.provinceId];
      if (province != null && !province.status.acceptsNewRides) return false;
      final countryId =
          (province != null && province.countryId.isNotEmpty)
              ? province.countryId
              : d.countryId;
      final country = _countriesById[countryId];
      if (country != null && !country.status.acceptsNewRides) return false;
      return true;
    }

    final activeDistricts = districts
        .where((d) => d.status.acceptsNewRides && parentActive(d))
        .toList();
    final activeSubs = subs
        .where(
          (s) =>
              s.status.acceptsNewRides &&
              s.supportsRide &&
              activeDistricts.any((d) => d.id == s.districtId),
        )
        .toList();
    _subById = {for (final s in activeSubs) s.id: s};

    _liveDistricts = [
      for (final d in activeDistricts)
        ServiceDistrictNode(
          id: d.id,
          nameAr: d.nameAr,
          nameEn: d.nameEn,
          customerVisible: d.customerVisible,
          subDistricts: [
            for (final s in activeSubs.where((x) => x.districtId == d.id))
              ServiceSubDistrictNode(
                id: s.id,
                nameAr: s.nameAr,
                nameEn: s.nameEn,
                center: s.center,
                searchRadiusKm: s.searchRadiusKm,
                boundary: s.boundary,
              ),
          ],
        ),
    ].where((d) => d.subDistricts.isNotEmpty).toList();

    _allDistricts = [
      for (final d in districts)
        ServiceDistrictNode(
          id: d.id,
          nameAr: d.nameAr,
          nameEn: d.nameEn,
          customerVisible: d.customerVisible,
          subDistricts: [
            for (final s in subs.where((x) => x.districtId == d.id))
              ServiceSubDistrictNode(
                id: s.id,
                nameAr: s.nameAr,
                nameEn: s.nameEn,
                center: s.center,
                searchRadiusKm: s.searchRadiusKm,
                boundary: s.boundary,
              ),
          ],
        ),
    ];
    notifyListeners();
  }

  List<ServiceProvince> get _seedProvinces => [
        ServiceProvince(
          id: _seedProvinceId,
          countryId: _seedCountryId,
          nameEn: BabilRegions.provinceNameEn,
          nameAr: BabilRegions.provinceNameAr,
        ),
      ];

  List<ServiceDistrict> get _seedDistrictsAsService => [
        for (final d in BabilRegions.seedDistricts)
          ServiceDistrict(
            id: d.id,
            provinceId: _seedProvinceId,
            countryId: _seedCountryId,
            nameEn: d.nameEn,
            nameAr: d.nameAr,
          ),
      ];

  List<ServiceSubDistrict> get _seedSubsAsService => [
        for (final d in BabilRegions.seedDistricts)
          for (final s in d.subDistricts)
            ServiceSubDistrict(
              id: s.id,
              districtId: d.id,
              provinceId: _seedProvinceId,
              countryId: _seedCountryId,
              nameEn: s.nameEn,
              nameAr: s.nameAr,
              center: s.center,
              searchRadiusKm: s.searchRadiusKm,
            ),
      ];

  List<BabilDistrict> _nodesToBabil(List<ServiceDistrictNode> nodes) {
    return [
      for (final d in nodes)
        BabilDistrict(
          id: d.id,
          nameAr: d.nameAr,
          nameEn: d.nameEn,
          subDistricts: [
            for (final s in d.subDistricts)
              BabilSubDistrict(
                id: s.id,
                nameAr: s.nameAr,
                nameEn: s.nameEn,
                center: s.center,
                searchRadiusKm: s.searchRadiusKm,
                boundary: s.boundary,
              ),
          ],
        ),
    ];
  }

  List<BabilDistrict> get districtsAsBabil {
    if (!_synced) {
      return BabilRegions.seedDistricts;
    }
    return _nodesToBabil(_liveDistricts ?? const <ServiceDistrictNode>[]);
  }

  /// Admin city filters: prefer all configured districts; never return empty.
  List<BabilDistrict> get districtsForAdminFilters {
    if (!_synced) {
      return BabilRegions.seedDistricts;
    }
    final all = _nodesToBabil(_allDistricts ?? const <ServiceDistrictNode>[]);
    if (all.isNotEmpty) return all;
    final live = districtsAsBabil;
    if (live.isNotEmpty) return live;
    return BabilRegions.seedDistricts;
  }

  List<BabilDistrict> get customerDistrictsAsBabil {
    if (!_synced) {
      return BabilRegions.seedCustomerDistricts;
    }
    final live = _liveDistricts ?? const <ServiceDistrictNode>[];
    final visible = live.where((d) => d.customerVisible).toList();
    final source = visible.isEmpty ? live : visible;
    return [
      for (final d in source)
        BabilDistrict(
          id: d.id,
          nameAr: d.nameAr,
          nameEn: d.nameEn,
          subDistricts: [
            for (final s in d.subDistricts)
              BabilSubDistrict(
                id: s.id,
                nameAr: s.nameAr,
                nameEn: s.nameEn,
                center: s.center,
                searchRadiusKm: s.searchRadiusKm,
                boundary: s.boundary,
              ),
          ],
        ),
    ];
  }

  ServiceSubDistrict? subDistrictConfig(String id) => _subById[id];

  /// All governorates (provinces) for admin filters — dynamic, never hardcoded.
  List<ServiceProvince> get provincesForAdminFilters {
    final list = _provincesById.values.toList();
    if (list.isEmpty) return _seedProvinces;
    list.sort((a, b) => a.nameEn.toLowerCase().compareTo(b.nameEn.toLowerCase()));
    return list;
  }

  /// Governorates offered to customers in the cascading area selector:
  /// active + customerVisible, with seed fallback so cold-start (before the
  /// first Firestore snapshot) still shows Babil. Never empty. New
  /// governorates (e.g. Karbala, Najaf) appear automatically once Admin
  /// activates them — no app update required.
  List<ServiceProvince> get customerProvinces {
    if (!_synced) return _seedProvinces;
    final all = _provincesById.values.toList();
    final visible = all.where((p) => p.isActive && p.customerVisible).toList();
    final list = visible.isEmpty ? all : visible;
    if (list.isEmpty) return _seedProvinces;
    list.sort((a, b) => a.nameEn.toLowerCase().compareTo(b.nameEn.toLowerCase()));
    return list;
  }

  /// Customer-visible districts (with their customer-visible sub-districts)
  /// under [provinceId], matching the same active/customerVisible filtering
  /// as [customerDistrictsAsBabil]. Backed entirely by live Firestore data
  /// once synced, so Admin-added districts appear without an app update.
  List<BabilDistrict> customerDistrictsForProvince(String provinceId) {
    final all = customerDistrictsAsBabil;
    if (provinceId.isEmpty) return all;
    if (!_synced) {
      return provinceId == _seedProvinceId ? all : const [];
    }
    final idsInProvince = _rawDistricts
        .where((d) => d.provinceId == provinceId)
        .map((d) => d.id)
        .toSet();
    return all.where((d) => idsInProvince.contains(d.id)).toList();
  }

  List<ServiceDistrict> get _allDistrictsForAdmin {
    if (_rawDistricts.isNotEmpty) return _rawDistricts;
    return _seedDistrictsAsService;
  }

  List<ServiceSubDistrict> get _allSubsForAdmin {
    if (_rawSubs.isNotEmpty) return _rawSubs;
    return _seedSubsAsService;
  }

  List<ServiceDistrict> districtsForProvince(String? provinceId) {
    final source = _allDistrictsForAdmin;
    if (provinceId == null || provinceId.isEmpty) return source;
    return source.where((d) => d.provinceId == provinceId).toList();
  }

  List<ServiceSubDistrict> subsForDistrict(String? districtId) {
    final source = _allSubsForAdmin;
    if (districtId == null || districtId.isEmpty) return source;
    return source.where((s) => s.districtId == districtId).toList();
  }

  String? provinceIdForDistrict(String districtId) {
    for (final d in _allDistrictsForAdmin) {
      if (d.id == districtId) return d.provinceId;
    }
    return _seedProvinceId;
  }

  ServiceProvince? provinceById(String id) {
    final p = _provincesById[id];
    if (p != null) return p;
    for (final seed in _seedProvinces) {
      if (seed.id == id) return seed;
    }
    return null;
  }

  ServiceDistrict? districtById(String id) {
    for (final d in _allDistrictsForAdmin) {
      if (d.id == id) return d;
    }
    return null;
  }

  String localizedProvinceName(String id, {required bool isAr}) {
    if (id == _seedProvinceId) {
      return isAr ? BabilRegions.provinceNameAr : BabilRegions.provinceNameEn;
    }
    final p = provinceById(id);
    if (p == null) return id;
    return isAr ? p.nameAr : p.nameEn;
  }

  String localizedDistrictName(String id, {required bool isAr}) {
    final d = districtById(id);
    if (d == null) {
      final seed = BabilRegions.districtById(id);
      return isAr ? seed.nameAr : seed.nameEn;
    }
    return isAr ? d.nameAr : d.nameEn;
  }

  String localizedSubName(String id, {required bool isAr}) {
    for (final s in _allSubsForAdmin) {
      if (s.id == id) return isAr ? s.nameAr : s.nameEn;
    }
    final live = _subById[id];
    if (live != null) return isAr ? live.nameAr : live.nameEn;
    for (final d in BabilRegions.seedDistricts) {
      for (final s in d.subDistricts) {
        if (s.id == id) return isAr ? s.nameAr : s.nameEn;
      }
    }
    return id;
  }

  /// True if [point] falls inside any active sub-district's effective
  /// boundary (Admin-drawn polygon, or one synthesized from center + radius).
  bool isWithinAnyActiveArea(LatLng point) {
    for (final sub in _subById.values) {
      if (GeoPolygon.isWithinBoundary(
        point: point,
        center: sub.center,
        radiusKm: sub.searchRadiusKm,
        storedBoundary: sub.boundary,
      )) {
        return true;
      }
    }
    // Before first sync, fall back to seed geometry so cold-start still works.
    if (!_synced) {
      for (final d in BabilRegions.seedDistricts) {
        for (final s in d.subDistricts) {
          if (GeoPolygon.isWithinBoundary(
            point: point,
            center: s.center,
            radiusKm: s.searchRadiusKm,
            storedBoundary: s.boundary,
          )) {
            return true;
          }
        }
      }
    }
    return false;
  }

  /// Validates a district/sub pair for a new ride request.
  /// Returns null when allowed, otherwise a short error code.
  String? validateForNewRide({
    required String districtId,
    required String subDistrictId,
    LatLng? pickup,
    LatLng? destination,
  }) {
    if (_synced) {
      final sub = _subById[subDistrictId];
      if (sub == null || sub.districtId != districtId) {
        return 'area_inactive';
      }
      if (!sub.acceptsNewRideRequests()) {
        return sub.operatingHours.alwaysOpen ||
                sub.operatingHours.isOpenAt(DateTime.now())
            ? 'area_inactive'
            : 'area_closed';
      }
      bool withinSub(LatLng point) {
        return GeoPolygon.isWithinBoundaryUnique(
          point: point,
          center: sub.center,
          radiusKm: sub.searchRadiusKm,
          storedBoundary: sub.boundary,
          others: [
            for (final other in _subById.values)
              if (other.id != sub.id)
                GeoArea(
                  center: other.center,
                  radiusKm: other.searchRadiusKm,
                  boundary: other.boundary,
                ),
          ],
        );
      }

      if (pickup != null && !withinSub(pickup)) {
        return 'outside_area';
      }
      if (destination != null && !withinSub(destination)) {
        return 'outside_area';
      }
      return null;
    }
    // Pre-sync: allow seed areas only.
    final seedOk = BabilRegions.seedDistricts.any(
      (d) =>
          d.id == districtId && d.subDistricts.any((s) => s.id == subDistrictId),
    );
    return seedOk ? null : 'area_inactive';
  }
}

File: hilla_ride_ios/HelloTukTuk/Core/Services/ServiceAreaCatalog.swift
import CoreLocation
import FirebaseFirestore
import Foundation

/// Lightweight governorate summary for the customer cascading area
/// selector (Governorate → District → Subdistrict). Iraq is fixed and never
/// shown; this only lists governorates such as Babil, Karbala, Najaf, etc.
struct ServiceProvinceSummary: Identifiable, Hashable {
    let id: String
    let nameEn: String
    let nameAr: String

    func displayName(language: AppLanguage) -> String {
        language == .arabic ? nameAr : nameEn
    }
}

/// Live Iraq service-area catalog (Province → District → Subdistrict).
/// After the first Firestore snapshot, seed data is never used for booking.
@MainActor
final class ServiceAreaCatalog: ObservableObject {
    static let shared = ServiceAreaCatalog()

    @Published private(set) var synced = false
    @Published private(set) var liveDistricts: [BabilDistrict] = []

    private var listeners: [ListenerRegistration] = []
    private var countries: [String: [String: Any]] = [:]
    private var provinces: [String: [String: Any]] = [:]
    private var districts: [String: [String: Any]] = [:]
    private var subs: [String: [String: Any]] = [:]

    private init() {}

    func start() {
        guard listeners.isEmpty else { return }
        let db = Firestore.firestore()
        listeners.append(db.collection("serviceCountries").addSnapshotListener { [weak self] snap, error in
            Task { @MainActor in
                guard let self else { return }
                if let error {
                    print("ServiceAreaCatalog serviceCountries error: \(error.localizedDescription)")
                    return
                }
                guard let snap else { return }
                self.countries = Dictionary(
                    uniqueKeysWithValues: snap.documents.map { ($0.documentID, $0.data()) }
                )
                self.rebuild()
            }
        })
        listeners.append(db.collection("serviceProvinces").addSnapshotListener { [weak self] snap, error in
            Task { @MainActor in
                guard let self else { return }
                if let error {
                    print("ServiceAreaCatalog serviceProvinces error: \(error.localizedDescription)")
                    return
                }
                guard let snap else { return }
                self.provinces = Dictionary(
                    uniqueKeysWithValues: snap.documents.map { ($0.documentID, $0.data()) }
                )
                self.rebuild()
            }
        })
        listeners.append(db.collection("serviceDistricts").addSnapshotListener { [weak self] snap, error in
            Task { @MainActor in
                guard let self else { return }
                if let error {
                    print("ServiceAreaCatalog serviceDistricts error: \(error.localizedDescription)")
                    return
                }
                guard let snap else { return }
                self.districts = Dictionary(
                    uniqueKeysWithValues: snap.documents.map { ($0.documentID, $0.data()) }
                )
                self.rebuild()
            }
        })
        listeners.append(db.collection("serviceSubDistricts").addSnapshotListener { [weak self] snap, error in
            Task { @MainActor in
                guard let self else { return }
                if let error {
                    print("ServiceAreaCatalog serviceSubDistricts error: \(error.localizedDescription)")
                    return
                }
                guard let snap else { return }
                self.subs = Dictionary(
                    uniqueKeysWithValues: snap.documents.map { ($0.documentID, $0.data()) }
                )
                self.rebuild()
            }
        })
    }

    var districtsForApps: [BabilDistrict] {
        if !synced { return BabilRegions.seedDistricts }
        return liveDistricts
    }

    var customerDistricts: [BabilDistrict] {
        if !synced { return BabilRegions.seedCustomerDistricts }
        let visible = liveDistricts.filter { district in
            if let data = districts[district.id] {
                return (data["customerVisible"] as? Bool) ?? true
            }
            return true
        }
        let list = visible.isEmpty ? liveDistricts : visible
        if list.isEmpty { return BabilRegions.seedCustomerDistricts }
        return list
    }

    /// Governorates offered to customers in the cascading area selector:
    /// active + customerVisible, with a Babil seed fallback before the
    /// first Firestore sync. Never empty. New governorates (e.g. Karbala,
    /// Najaf) appear automatically once Admin activates them — no app
    /// update required.
    var customerProvinces: [ServiceProvinceSummary] {
        let seed = [
            ServiceProvinceSummary(
                id: BabilRegions.seedProvinceId,
                nameEn: BabilRegions.provinceNameEn,
                nameAr: BabilRegions.provinceNameAr
            )
        ]
        if !synced { return seed }
        let all = provinces.map { id, data in
            ServiceProvinceSummary(
                id: id,
                nameEn: data["nameEn"] as? String ?? id,
                nameAr: data["nameAr"] as? String ?? id
            )
        }
        let visible = all.filter { p in
            guard let data = provinces[p.id] else { return true }
            let customerVisible = (data["customerVisible"] as? Bool) ?? true
            return customerVisible && accepts(data["status"] as? String)
        }
        let list = visible.isEmpty ? all : visible
        let result = list.isEmpty ? seed : list
        return result.sorted { $0.nameEn < $1.nameEn }
    }

    /// Customer-visible districts (with their sub-districts) under
    /// `provinceId`, matching the same active/customerVisible filtering as
    /// `customerDistricts`. Backed entirely by live Firestore data once
    /// synced, so Admin-added districts appear without an app update.
    func districtsForProvince(_ provinceId: String) -> [BabilDistrict] {
        let all = customerDistricts
        if provinceId.isEmpty { return all }
        if !synced {
            return provinceId == BabilRegions.seedProvinceId ? all : []
        }
        let inProvince = all.filter { (districts[$0.id]?["provinceId"] as? String) == provinceId }
        if inProvince.isEmpty, provinceId == BabilRegions.seedProvinceId {
            return all
        }
        return inProvince
    }

    /// Governorate id that owns `districtId`, falling back to the seed
    /// province before first sync or when the district is unknown.
    func provinceIdForDistrict(_ districtId: String) -> String {
        if let provinceId = districts[districtId]?["provinceId"] as? String, !provinceId.isEmpty {
            return provinceId
        }
        return BabilRegions.seedProvinceId
    }

    /// True if `point` falls inside any active sub-district's effective
    /// boundary (Admin-drawn polygon, or one synthesized from center + radius).
    func isWithinAnyActiveArea(_ point: CLLocationCoordinate2D) -> Bool {
        for district in districtsForApps {
            for sub in district.subDistricts {
                if GeoPolygon.isWithinBoundary(
                    point: point,
                    center: sub.center,
                    radiusKm: sub.searchRadiusKm,
                    storedBoundary: sub.boundary
                ) {
                    return true
                }
            }
        }
        return false
    }

    func validateForNewRide(
        districtId: String,
        subDistrictId: String,
        pickup: CLLocationCoordinate2D? = nil,
        destination: CLLocationCoordinate2D? = nil
    ) -> String? {
        if !synced {
            let ok = BabilRegions.seedDistricts.contains {
                $0.id == districtId && $0.subDistricts.contains { $0.id == subDistrictId }
            }
            return ok ? nil : "area_inactive"
        }
        guard let district = liveDistricts.first(where: { $0.id == districtId }),
              let sub = district.subDistricts.first(where: { $0.id == subDistrictId })
        else {
            return "area_inactive"
        }
        let others: [GeoArea] = liveDistricts.flatMap { d in
            d.subDistricts.compactMap { other -> GeoArea? in
                guard other.id != sub.id else { return nil }
                return GeoArea(center: other.center, radiusKm: other.searchRadiusKm, boundary: other.boundary)
            }
        }
        func withinSub(_ point: CLLocationCoordinate2D) -> Bool {
            GeoPolygon.isWithinBoundaryUnique(
                point: point,
                center: sub.center,
                radiusKm: sub.searchRadiusKm,
                storedBoundary: sub.boundary,
                others: others
            )
        }
        if let pickup, !withinSub(pickup) {
            return "outside_area"
        }
        if let destination, !withinSub(destination) {
            return "outside_area"
        }
        return nil
    }

    /// Parses a Firestore `boundary` field (array of `{lat, lng}` maps) into
    /// a polygon, or nil when absent/invalid (< 3 points).
    private static func parseBoundary(_ raw: Any?) -> [CLLocationCoordinate2D]? {
        guard let list = raw as? [[String: Any]] else { return nil }
        let points: [CLLocationCoordinate2D] = list.compactMap { entry in
            guard let lat = (entry["lat"] as? NSNumber)?.doubleValue,
                  let lng = (entry["lng"] as? NSNumber)?.doubleValue
            else { return nil }
            return CLLocationCoordinate2D(latitude: lat, longitude: lng)
        }
        return points.count >= 3 ? points : nil
    }

    /// True when the governorate/province is staged for customers (defaults
    /// to visible so existing provinces without the field are unaffected).
    func provinceCustomerVisible(_ provinceId: String) -> Bool {
        guard let data = provinces[provinceId] else { return true }
        return (data["customerVisible"] as? Bool) ?? true
    }

    private func accepts(_ status: String?) -> Bool {
        (status ?? "inactive") == "active"
    }

    private func supportsRide(_ services: [String]) -> Bool {
        services.contains { $0.lowercased() == "ride" }
    }

    private func rebuild() {
        synced = true
        var built: [BabilDistrict] = []
        for (districtId, d) in districts {
            guard accepts(d["status"] as? String) else { continue }
            let provinceId = d["provinceId"] as? String ?? ""
            if let province = provinces[provinceId], !accepts(province["status"] as? String) {
                continue
            }
            let countryId = (provinces[provinceId]?["countryId"] as? String)
                ?? (d["countryId"] as? String)
                ?? ""
            if let country = countries[countryId], !accepts(country["status"] as? String) {
                continue
            }

            let subNodes: [BabilSubDistrict] = subs.compactMap { subId, s in
                guard (s["districtId"] as? String) == districtId else { return nil }
                guard accepts(s["status"] as? String) else { return nil }
                let services = (s["services"] as? [String]) ?? ["ride"]
                guard supportsRide(services) else { return nil }
                let lat = (s["latitude"] as? NSNumber)?.doubleValue ?? 0
                let lng = (s["longitude"] as? NSNumber)?.doubleValue ?? 0
                let radius = (s["searchRadiusKm"] as? NSNumber)?.doubleValue
                    ?? BabilRegions.defaultSubDistrictRadiusKm
                return BabilSubDistrict(
                    id: subId,
                    nameEn: s["nameEn"] as? String ?? subId,
                    nameAr: s["nameAr"] as? String ?? subId,
                    center: CLLocationCoordinate2D(latitude: lat, longitude: lng),
                    searchRadiusKm: radius,
                    boundary: Self.parseBoundary(s["boundary"])
                )
            }
            .sorted { $0.nameEn < $1.nameEn }

            guard !subNodes.isEmpty else { continue }
            built.append(
                BabilDistrict(
                    id: districtId,
                    nameEn: d["nameEn"] as? String ?? districtId,
                    nameAr: d["nameAr"] as? String ?? districtId,
                    subDistricts: subNodes
                )
            )
        }
        liveDistricts = built.sorted { $0.nameEn < $1.nameEn }
    }
}
